### ISSUE TITLE
UnsupportedOperationException when method explicitly called with null

### ISSUE THREAD

**Issue by sgalles** [ceylon/ceylon-compiler#1985]
This code
```ceylon
shared void run(){
        void foo(void bar(Integer? b)){
            bar(null);
        } 
        foo(void(Integer? b){print(b);});
}
```
When compiled, prints a strange warning :
```
warning: non-varargs call of varargs method with inexact argument type for last parameter;
            bar(null);
                ^
  cast to Object for a varargs call
  cast to Object[] for a non-varargs call and to suppress this warning
```
And when executed :
```ceylon
ceylon run: null
java.lang.UnsupportedOperationException
        at com.redhat.ceylon.compiler.java.language.AbstractCallable.$call$(AbstractCallable.java:79)
        at foo.run_$1foo_.foo(run.ceylon:26)
        at foo.run_.run(run.ceylon:28)
        at foo.run_.main(run.ceylon)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at ceylon.modules.api.runtime.SecurityActions.invokeRunInternal(SecurityActions.java:58)
        at ceylon.modules.api.runtime.SecurityActions.invokeRun(SecurityActions.java:48)
        at ceylon.modules.api.runtime.AbstractRuntime.invokeRun(AbstractRuntime.java:101)
        at ceylon.modules.api.runtime.AbstractRuntime.execute(AbstractRuntime.java:167)
        at ceylon.modules.api.runtime.AbstractRuntime.execute(AbstractRuntime.java:151)
        at ceylon.modules.Main.execute(Main.java:69)
        at ceylon.modules.Main.main(Main.java:42)
```

**Comment by tombentley** [comment]
The warning message is complaining that the invocation is ambiguous between a varargs call to `$call$(Object...)` passing a single null argument and an array call to `$call$(Object...)` passing a null array. I think that's just bad compiler diagnostics though. The "real" ambiguity seems to be that we could be calling either `$call$(Object...)` or `$call$(Object)`. We intend the latter, but the compiler end up emittting a call to the former. (My brief reading of http://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.12.2.5 is that this should be considered as ambiguous, because that doesn't define the "more specific" relation in the case of variable arity versus fixed arity).

In practice it seems that a typecast of the first argument to `Object` is sufficient to get rid of the warning and make the compiler select `$call$(Object)`. The problem is that nothing in the code generator really knows what's going on here. The `Invocation` doesn't know it's invoking on a `Callable`, and so when we come to transform arguments we've no way to know we need a typecast. But I can hack around that.

Then it transpires there's a similar problem with functional parameters with variadic parameter lists (no great surprise).

Then another related problem with `Util.sequentialCopy()` (being called from the code generated for the Callable) being passed a `null` argument, and that being resolved as a non-varargs call passing a null array. The obvious thing to do here is to pass an explicit array rather than letting javac box it for us. But `Util.sequentialCopy()` is a generic method and the variadic parameter was a type parameter, which meant we've been using the inferred (or explicit) type to construct the array, so we can't now just box the arguments into an `Object[]`, we need to know the type of the type argument and make an array of that type.

Now I'm starting to think it's a mistake to let `Util.sequentialCopy()` take a `T...` and it should at least be a `Object...` (I don't think that would break BC between 1.1 and 1.2). Ideally it would be an `Object[]` to eliminate ambiguities in the future (but that would definitely break BC).

I think we need to be a *lot* more careful about where we're using Java variadic parameters in the runtime: If they're called from generated code there's the possibility for ambiguity.

**Comment by FroMage** [comment]
Don't we know what sort of thing we're invoking, though? Isn't that done in `IndirectInvocation`? We have a similar automatic cast (actually I think we always do the array packing manually) for invoking Java variadic methods.

**Comment by tombentley** [comment]
This isn't using `IndirectInvocation` because indirect-ness is when we're invoking a *value* of `Callable` type. In this case there's a `Method` model available at compile time so it's treated as a normal positional invocation (it just happens that the thing we're calling is a Callable because we can't reify a functional parameter as a Java method).

**Comment by tombentley** [comment]
This should now be fixed, but I've opened an issue against ceylon.language to review all the existing uses of `...` because I'd be amazed if there weren't more problems like this one.
